libbeat/common/kubernetes/util.go

```diff
@@ -94,54 +93,65 @@ func IsInCluster(kubeconfig string) bool {
 // If host is provided in the config use it directly.
 // If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
+// If node can not be discovered return NODE_NAME env var as default value. In case it is not set return error.
+func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (string, error) {
 	if host != "" {
 		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+		return host, nil
+	}
+	nodeNameEnv := os.Getenv("NODE_NAME")
+	var envError, logerror error
+	if nodeNameEnv == "" {
+		envError = errors.New("kubernetes: NODE_NAME environment variable was not set")
 	}
 	ctx := context.TODO()
 	if inCluster {
 		ns, err := InClusterNamespace()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, enrichedError([]error{envError, logerror})
 		}
 		podName, err := os.Hostname()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, enrichedError([]error{envError, logerror})
 		}
 		log.Infof("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
 		pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
 		if err != nil {
-			log.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
+			log.Error(logerror)
+			return nodeNameEnv, enrichedError([]error{envError, logerror})
```

Comment:
Maybe `Wrap` can be used here instead of `enrichedError `? See https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/x-pack/metricbeat/module/stan/stats/stats.go#L73

Reply:
Good catch. It does almost the same thing